uart/gps_uart/gps_uart.c

```diff
@@ -0,0 +1,251 @@
+ *
+ * @param nmea_string The NMEA sentence to validate.
+ * @return true if the checksum is valid, false otherwise.
+ */
+bool validate_nmea_checksum(char *nmea_string)
+{
+    int len = strlen(nmea_string);
+    printf("Checksum Validation - String Length: %d\n", len);
+    printf("Full NMEA String: %s", nmea_string);
+
+    if (len < 7)
+    {
+        printf("Invalid: Too short (< 7 chars)\n");
+        return false;
+    }
+
+    char *checksum_ptr = strchr(nmea_string, '*');
+    if (!checksum_ptr)
+    {
+        printf("Invalid: No checksum marker (*) found\n");
+        return false;
+    }
+
+    uint8_t calculated_checksum = 0;
+    for (char *p = nmea_string + 1; p < checksum_ptr; p++)
+    {
+        calculated_checksum ^= *p;
+    }
+
+    char hex_checksum[3];
+    snprintf(hex_checksum, sizeof(hex_checksum), "%02X", calculated_checksum);
+
+    printf("Calculated Checksum: %s\n", hex_checksum);
+    printf("Received Checksum: %s\n", checksum_ptr + 1);
+
+    bool is_valid = strncmp(hex_checksum, checksum_ptr + 1, 2) == 0;
+    printf("Checksum Validation: %s\n", is_valid ? "VALID" : "INVALID");
+
+    return is_valid;
+}
+
+
+/**
+ * @brief Converts an NMEA coordinate to decimal degrees.
+ *
+ * This function converts an NMEA-formatted coordinate (DDMM.MMMM) to
+ * decimal degrees (DD.DDDDDD).
+ *
+ * @param nmea_coord The NMEA coordinate string to convert.
+ * @param decimal_coord Pointer to store the converted decimal coordinate.
+ */
+void convert_nmea_to_decimal(char *nmea_coord, float *decimal_coord)
+{
+    float degrees = atof(nmea_coord) / 100.0;
+    int int_degrees = (int)degrees;
+    float minutes = (degrees - int_degrees) * 100.0;
+
+    *decimal_coord = int_degrees + (minutes / 60.0);
+}
+
+
+/**
+ * @brief Parses an NMEA sentence and extracts GPS data.
+ *
+ * This function parses an NMEA sentence (currently only $GPRMC) and extracts
+ * latitude, longitude, and validity information. It also validates the checksum.
+ *
+ * @param nmea_string The NMEA sentence to parse.
+ * @param gps_data Pointer to the GPSData structure to store the parsed data.
+ * @return true if the sentence was successfully parsed, false otherwise.
+ */
+bool parse_nmea_gps(char *nmea_string, GPSData *gps_data)
+{
+    if (!validate_nmea_checksum(nmea_string))
+    {
+        printf("Invalid NMEA Checksum\n");
+        return false;
+    }
+
+    // Ignore all NMEA strings that are not of type GPRMC
+    if (strncmp(nmea_string, "$GPRMC", 6) != 0)
+    {
+        return false;
+    }
+    
+    char *token;
+    char *tokens[12] = {0};
+    int token_count = 0;
+
+    // Tokenize the string
+    token = strtok(nmea_string, ",");
+    while (token != NULL && token_count < 12)
+    {
+        tokens[token_count++] = token;
+
+        token = strtok(NULL, ",");
+    }
+
+    // Check if we have enough tokens and data is valid
+    if (token_count >= 12 && strcmp(tokens[2], "A") == 0)
```

Comment:
This fails for this string "$GPRMC,193045.000,A,5214.2340,N,00008.7338,E,0.00,0.00,130225,,,D*66"